Validate resource

pradeepitm12 opened the pull request:
Jenkins throws null pointer exception if any field is mising in resource file.

This patch validates the resource file before processing it.

- Give option to validate template by default validates.
- For an invalid template show proper message and stop jenkins processing.
- Readme Update for loadResource method arguments.

Resolves : https://github.com/fabric8io/osio-pipeline/issues/52

sthaha commented on vars/loadResources.groovy:
```diff
@@ -1,17 +1,24 @@
   }
 
   def yaml = readYaml(file: args.file)
+  def validate = args.validate ?: "true"
```
`true` instead of `"true"` . Groovy has support of boolean just like Java

pradeepitm12 replied:
changed to boolean

sthaha commented on vars/loadResources.groovy:
```diff
@@ -1,17 +1,24 @@
-  // for each of the above, loadResources must return [kind: [Resources, ...]]
+  // for each of the akbove, loadResources must return [kind: [Resources, ...]]
+
+  if (validate.toLowerCase() == "true") {
```
`if (validate) {` would suffice.

pradeepitm12 replied:
this is handled by boolean type

sthaha commented on vars/loadResources.groovy:
```diff
@@ -1,17 +1,24 @@
+  // for each of the akbove, loadResources must return [kind: [Resources, ...]]
+
+  if (validate.toLowerCase() == "true") {
+    Utils.shWithOutput(this, "oc apply --dry-run=true --validate=true -f $args.file -o yaml");
```
why -o yaml?

sthaha commented on vars/loadResources.groovy:
```diff
@@ -1,17 +1,24 @@
   }
 
   def yaml = readYaml(file: args.file)
+  def validate = args.validate ?: "true"
```
have you validated that this wouldn't set `validate` to `true` even if it is set to `false`?

hrishin replied:
yeah issue is if `args.validate` is set to `false` then `args.validate ?: true` is always evaluates to `true`

better way make it as `def validate = args.validate == false ? false : true` if we want to use boolean type

pradeepitm12 replied:
@sthaha Yes I test it. But now as we have changed variable to boolean, it works fine.
@hrishin thanks.

hrishin commented on vars/loadResources.groovy:
```diff
@@ -1,17 +1,24 @@
   //    and items property is a list of resources
   //
-  // for each of the above, loadResources must return [kind: [Resources, ...]]
+  // for each of the akbove, loadResources must return [kind: [Resources, ...]]
```
`akbove` -> `above`

pradeepitm12 replied:
Typo, changed.

pradeepitm12 commented:
@hrishin @sthaha 
I have incorporated all the comments, squashed multiple commits to one, removed usless file.
It's a kind request to review the changes.

piyush-garg commented on .gitignore:
```diff
@@ -0,0 +1 @@
+.idea/*
```
remove this one

pradeepitm12 replied:
Done

hrishin commented on vars/loadResources.groovy:
```diff
@@ -1,9 +1,12 @@
     error "Missing manadatory parameter: file"
   }
 
   def yaml = readYaml(file: args.file)
```
@pradeepitm12 would it be better to move validation call before `readYaml`?

pradeepitm12 replied:
Yes, we can validate first and then load it.
